### PR TITLE
Further adjustment to airhose movement

### PR DESCRIPTION
**Pull request by peternewell**
https://blueprints.launchpad.net/or/+spec/advanced-coupler-refinements 

Further refinements to advanced coupling.

### CODE REVIEW

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs [hunk 1]
```diff
@@ -1303,31 +1312,38 @@ public void UpdateTrainDerailmentRisk()
                     CarBehind.FrontAirHoseHeightAdjustmentM = frontairhoseheightadjustmentreferenceM - CarBehind.FrontAirHoseHeightAdjustmentM;
                 }
 
-                var rearairhoseangleadjustmentreferenceRad = (float)Math.Cos(CouplerDistanceThisCarM / CarAirHoseLengthM);
-                var frontairhoseangleadjustmentreferenceRad = (float)Math.Cos(CouplerDistanceBehindCarM / CarAirHoseLengthM);
+                // Calculate angle adjustments
+                var rearairhoseangleadjustmentreferenceRad = (float)Math.Asin(CarAirHoseHorizontalLengthM / CarAirHoseLengthM);
```
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1316]:
```suggestion
                var rearAirHoseAngleAdjustmentReferenceRad = (float)Math.Asin(CarAirHoseHorizontalLengthM / CarAirHoseLengthM);
```

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs [hunk 2]
```diff
@@ -1303,31 +1312,38 @@ public void UpdateTrainDerailmentRisk()
...
+                var frontairhoseangleadjustmentreferenceRad = (float)Math.Asin(CarBehind.CarAirHoseHorizontalLengthM / CarAirHoseLengthM);
```
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1317]:
```suggestion
                var frontAirHoseAngleAdjustmentReferenceRad = (float)Math.Asin(CarBehind.CarAirHoseHorizontalLengthM / CarAirHoseLengthM);
```

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSWagon.cs [hunk 3]
```diff
@@ -472,14 +472,9 @@ public virtual void LoadFromWagFile(string wagFilePath)
...
-            if (CarAirHoseLengthM < couplerlength)
+            if (CarAirHoseHorizontalLengthM == 0)
             {
-                CarCouplerFaceLengthM = CarBodyLengthM + (0.4f * 2.0f); // Assume a coupler length of 400mm at each end and add to car body length
-
-                if (Simulator.Settings.VerboseConfigurationMessages)
-                {
-                    Trace.TraceInformation("Coupler length exceeded brake air hose length, so  ORTSLengthCouplerFace decreased to {0}", CarCouplerFaceLengthM);
-                }
+                CarAirHoseHorizontalLengthM = 0.3862f; // 15.2 inches
```
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/MSTSWagon.cs:477]: Could you extend the comment to link to the source for the 15.2" dimension? - E.g. a forum thread?

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs [hunk 4]
```diff
@@ -1277,21 +1281,26 @@ public void UpdateTrainDerailmentRisk()
...
                 }
 
-                // Calculate airhose angles and height
-                var rearairhoseheightadjustmentreferenceM = (float)Math.Sqrt((float)Math.Pow(CarAirHoseLengthM, 2) - (float)Math.Pow(CouplerDistanceThisCarM, 2));
-                var frontairhoseheightadjustmentreferenceM = (float)Math.Sqrt((float)Math.Pow(CarAirHoseLengthM, 2) - (float)Math.Pow(CouplerDistanceBehindCarM, 2));
+                // Calculate airhose angles and height adjustment values for the air hose.  Firstly the "rest point" is calculated, and then the real time point. 
+                // The height and angle variation are then calculated against "at rest" reference point. The air hose angle is used to rotate the hose in two directions, ie the Y and Z axis. 
+
+                // Calculate height adjustment.
+                var rearairhoseheightadjustmentreferenceM = (float)Math.Sqrt((float)Math.Pow(CarAirHoseLengthM, 2) - (float)Math.Pow(CarAirHoseHorizontalLengthM, 2));
```
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1288]:
```suggestion
                var rearAirHoseHeightAdjustmentReferenceM = (float)Math.Sqrt((float)Math.Pow(CarAirHoseLengthM, 2) - (float)Math.Pow(CarAirHoseHorizontalLengthM, 2));
```

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs [hunk 5]
```diff
@@ -1277,21 +1281,26 @@ public void UpdateTrainDerailmentRisk()
...
+                var frontairhoseheightadjustmentreferenceM = (float)Math.Sqrt((float)Math.Pow(CarAirHoseLengthM, 2) - (float)Math.Pow(CarBehind.CarAirHoseHorizontalLengthM, 2));
```
Comment (cjakeman) [inline Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1289]:
```suggestion
                var frontAirHoseHeightAdjustmentReferenceM = (float)Math.Sqrt((float)Math.Pow(CarAirHoseLengthM, 2) - (float)Math.Pow(CarBehind.CarAirHoseHorizontalLengthM, 2));
```

**Review by cjakeman — CHANGES_REQUESTED** [review]
Hi Peter,

This looks good and should add interesting visuals.

We're encouraging devs to get their Blueprints approved before the work is started, so it would help if you completed your Blueprint as an example.

A blueprint should declare what impact the new feature will have. Perhaps the simplest questions to cover are:

- Who will be affected? (E.g. all users, some users, devs?)
- How will they be affected? (E.g. new option/control/change to content files?)
- What risks will there be? (E.g. any other code affected, performance implications?)

Would you also link your PR to the forum thread.

Thanks,

Chris

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs (hunk 1 above)
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1316]: Chris,

Thanks for the feedback.

Blueprint updated as suggested.

Comment (cjakeman) [reply Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1316]: Hi Peter,

Thanks. Your blueprint is much better - good example for others.

Chris

**Review by cjakeman — COMMENTED** [review]
Hi Peter,

New features will be blocked in a few days time for v1.4 and will go into v1.5, so if you want to adopt my suggested changes, please do so quickly.
If not, let me know so I can approve as it currently is.

Thanks,

Chris

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs (hunk 5 above)
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1289]: Changed as requested

File: Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs (hunk 4 above)
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/TrainCar.cs:1288]: Changed as requested

File: Source/Orts.Simulation/Simulation/RollingStocks/MSTSWagon.cs (hunk 3 above)
Comment (peternewell) [reply Source/Orts.Simulation/Simulation/RollingStocks/MSTSWagon.cs:477]: I don't have a solid reference for this. 

It was done to suit the airhoses used in the demonstration.

**Review by cjakeman — APPROVED** [review]
Hi Peter,

Approved with thanks,

Chris

**Pull request merged**